rbd-mirror: adjust journal fetch properties based on memory target

trociny opened the pull request:
<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

trociny commented:
@dillaman I am not sure about the new configuration option name (and its description).

Also, what do you think about how this should interact with the current option `rbd_mirror_journal_max_fetch_bytes`? In my implementation, the `rbd_mirror_journal_max_fetch_bytes` is used as the minimum value the auto-tuner may set, though right now the special case when `rbd_mirror_journal_max_fetch_bytes` is set to 0 (which previously meant "fetch the whole object") is ignored (the value calculated by the auto-tuner will be used). I could change this that if `rbd_mirror_journal_max_fetch_bytes` is 0, we always fetch the whole object regardless of the auto-tuner calculations.

dillaman commented on src/common/options.cc:
```diff
@@ -7325,6 +7325,10 @@ static std::vector<Option> get_rbd_mirror_options() {
     .set_default(32768)
     .set_description("maximum bytes to read from each journal data object per fetch"),
 
+    Option("rbd_mirror_total_journal_max_fetch_bytes", Option::TYPE_UINT, Option::LEVEL_ADVANCED)
```
I'd prefer that we mimic the ```XYZ_memory_target``` options for OSD and (future) MON.

dillaman commented:
... and I'd vote that we remove the old ```rbd_mirror_journal_max_fetch_bytes``` option since it was a simple attempt to achieve a max memory usage.

trociny commented:
@dillaman I pushed the update (badly tested and with bad defaults) to discuss the questions I have:

1) Does it look like what you meant?
2) If it does then currently it fails here [1] when the number of mirrored images is more than 10. It seems the `PriorityCache` implementation does not assume so many caches. Also, looking at `get_chunk`, it seems it returns too large chunks for our case. I suppose I would need to add some  additional tuning parameters to `PriorityCache` instead of hard coded assumptions to work for our case.
3) Probably I have added too many "rbd_mirror_memory" configuration options, and we could use just some hard coded values for some of them (like "rbd_mirror_memory_cache_min").
4) If you think it was a good idea to introduce "rbd_mirror_memory_autotune" to be able to disable autotune, then it is not clear what fetch values to use in this case. Right now it would fetch the whole object. We could fallback to "rbd_mirror_journal_max_fetch_bytes" but you suggested to remove this option (it is not implemented yet).

[1] https://github.com/ceph/ceph/blob/master/src/common/PriorityCache.cc#L265

dillaman commented:
> 1. Does it look like what you meant?

From a quick review, yes.

> 2. If it does then currently it fails here [1] when the number of mirrored images is more than 10. It seems the `PriorityCache` implementation does not assume so many caches. Also, looking at `get_chunk`, it seems it returns too large chunks for our case. I suppose I would need to add some  additional tuning parameters to `PriorityCache` instead of hard coded assumptions to work for our case.

Hmm -- yeah, it looks like it the chunk size is going to be a minimum of around 300MiB. I wanted to re-use the priority cache because it's already added all the hooks for poking tcmalloc and balancing multiple caches. @markhpc, would it make sense to pull some of these RocksDB / OSD-specific tweaks out of the base priority cache?

> 3. Probably I have added too many "rbd_mirror_memory" configuration options, and we could use just some hard coded values for some of them (like "rbd_mirror_memory_cache_min").

They are at dev level so it's probably just as reasonable to use config options to avoid having magic numbers in the code. The "min" is probably something we can internally calculate, however, based on the number of images * splay width of the journal * journal object size. 

> 4. If you think it was a good idea to introduce "rbd_mirror_memory_autotune" to be able to disable autotune, then it is not clear what fetch values to use in this case. Right now it would fetch the whole object. We could fallback to "rbd_mirror_journal_max_fetch_bytes" but you suggested to remove this option (it is not implemented yet).

Really what we need is a way to say "enforce memory limits = true/false". With Rook and containerized deployments of Ceph, rbd-mirror is going to need to do a better job of handling its memory growth or else it's just going to get killed by cgroups when a malloc fails. If that means that it won't be able to potentially mirror more than X images, I think that's fine as long as we output an alert saying that it's overloaded for memory. If ImageReplayer attempts to add itself to the priority cache manager and there is no space available, the replay can fail out w/ "not enough memory".

trociny commented:
Updated. I fixed the issue with unnecessary extra allocations by adding a new virtual method `wants_extra` to `PriCache`. The problem could be solved in many other ways, I am ok with any if you don't like this one.

dillaman commented on src/common/PriorityCache.h:
```diff
@@ -98,6 +98,9 @@ namespace PriorityCache {
     virtual void set_cache_ratio(double ratio) = 0;
 
+    // If cache wants to get a little extra memory
+    virtual bool wants_extra() const = 0;
```
Might just be easier to have a constructor param on ```PriorityCache::Manager``` to enable/disable this functionality. That would avoid the need to iterate through hundreds or thousands of journal player caches on rebalance.

trociny replied:
@dillaman In general case I can easily imagine a mix of caches which want a different behavior here. And I think that instead of this method we might event want a method that would calculate how much this particular cache wants. And I don't think iterating here will have any measurable effect to total time of rebalance.

On the other hand taking that this interface may still be change in the future I am ok with your suggestion too.

dillaman commented on src/journal/JournalPlayer.cc:
```diff
@@ -44,16 +46,89 @@ struct C_HandleEntriesAvailable : public Context {
+
+  int64_t request_cache_bytes(PriorityCache::Priority pri,
+                              uint64_t total_cache) const override {
+    if (pri != PriorityCache::Priority::PRI0) {
```
Nit: perhaps just add ```ceph_assert(pri == PriorityCache::Priority::PRI0)``` since it seems like the other priorities aren't used / supported.

dillaman commented on src/common/options.cc:
```diff
@@ -7462,6 +7462,42 @@ static std::vector<Option> get_rbd_mirror_options() {
+    Option("rbd_mirror_memory_base", Option::TYPE_SIZE, Option::LEVEL_DEV)
+    .set_default(768_M)
+    .add_see_also("rbd_mirror_memory_autotune")
+    .set_description("When tcmalloc and cache autotuning is enabled, estimate the minimum amount of memory in bytes the OSD will need."),
```
Nit: OSD

trociny commented:
@dillaman updated, though I realized there are at least two problems with the current implementation:
1) `Manager::insert` and `erasure` methods have no protection against concurrent call, so the way how they  are called now in `JournalPlayer` is not safe.
2) at the moment when `m_pri_cache->get_cache_bytes` is called in `JournalPlayer`, the cache is just inserted and rebalance had not been running yet, so it will return 0 and the replay will fail.

To resolve this I am thinking of introducing a cache handler something like this:
```
// journal/Types.h:
struct CacheHandler {
  virtual void create(const std::string &cache_name) = 0;
  virtual void destroy(const std::string &cache_name) = 0;
  virtual uint64_t get_bytes(const std::string &cache_name) = 0;
}
```
The handler would be created by `Mirror` (which has control on cache manager) and passed to the `JournalPlayer`. The `create` method would call the cache manager's insert and rebalance methods protected by lock. The problem with this though that we would do too much rebalance when a mirroring pool is scanned for the first time and many images are added.

Then probably `create` method should be:
```
virtual void create(const std::string &cache_name, Context *on_finish);
```
and `on_finish` context is completed  when a scheduled periodic rebalance is complete. Or it might be easier (for `JournalPlayer`) to do this in `get_bytes`...

What do you think?

dillaman commented:
@trociny Would ```get_bytes``` also take a min size? How would it potentially reclaim bytes from the cache when new image replayers are added (e.g. in the future we might cache N * <splay-width*max-object-size> object players if we have enough memory in order to speed-up images that are behind). 

Perhaps ```get_bytes``` is really a ```minimum_bytes_factor``` that is a param passed in the ```create``` call. The ```create``` call could also take an interface class pointer w/ a ```set_available_bytes(uint64_t bytes, Context* on_finish)``` callback that is invoked on all journal replayers when space is needed / becomes available.

trociny commented:
> @trociny Would `get_bytes` also take a min size? How would it potentially reclaim bytes from the cache when new image replayers are added (e.g. in the future we might cache N * <splay-width*max-object-size> object players if we have enough memory in order to speed-up images that are behind).

My idea was the following. I supposed (correct me if I was wrong here) that `pri_cache->get_cache_bytes` would return the amount of bytes the cache manager reserved for this cache (after rebalance). The `JournalPlayer` would use this value when creating an `ObjectPlayer`. Note, when an object has been replayed and a new `ObjectPlayer` is created for the next object, `pri_cache->get_cache_bytes` is called again and the up to date `max_fetch_bytes` is used. So the bytes would be reclaimed to the cache every time when an object has been replayed.

dillaman commented:
@trociny What if an ```ObjectPlayer``` isn't making immediate forward progress? I think you still need the feedback loop between a rebalance saying "I need to steal previously allocated memory from image X to give to image Y" and the replayer actually releasing the memory. I think we also need to cover the case where a given object has an event whose size is greater than the provided ```max_fetch_bytes``` -- perhaps it needs to temporarily borrow up to ```2 * max object size``` capacity from a higher priority level to ensure forward progress.

trociny commented:
@dillaman I am not quite sure we really need this complication if we still want to use `PriorityCache::Manager`.

As I see the `PriorityCache::Manager` does not allocate memory but just gives an advice to its caches how much they could use taking into consideration the current usage and the caches wishes. Also if an `ObjectPlayer` is created with some `max_fetch_bytes` it does not mean that it will really use so much bytes. It is just its high limit. I don't see a problem if an `ObjectPlayer` created with high `max_fetch_bytes`  is not doing immediate forward progress and is still using this high value -- it does not prevent the Manager to decrease its cache size value during rebalance if it needs it for some other cache. The slowly progressing `ObjectPlayer` will still be using the outdated (overestimated) `max_fetch_bytes` value, but taking that it is not a real allocation but just a limit I don't expect it will cause real overallocation. And I suppose a slowly progressing `ObjectPlayer` usually means that it actually has little to do, so it is unlikely it really allocates much for fetch.

Also note that `PriorityCache::Manager` does not react immediately it just does its recalculations periodically.

Anyway, I could make `ObjectPlayer`'s `max_fetch_bytes` dynamic, by passing the priority cache (or some handler) to the `ObjectPlayer`, so it would calculate the actual `max_fetch_bytes` before every object read. I think it would work as your feedback loop, but would be much simpler. Still I am not sure we need even this.

dillaman commented:
@trociny Yes, understood that the ```PriorityCache::Manager``` doesn't actually allocate memory. Right now, it's pretty hard-coded that we just want to ensure enough memory to read <splay width> * <object size> * <number of replayers>. What I am getting at is that (1) it might need up to 2 x <object size>, and (2) in the future, it would be great if we had a way to provide more memory to the ```JournalReplayer``` so that it could pre-fetch additional objects (the original version used to pre-fetch two full object sets but I had to remove it due to possible memory constraints with hundreds of images being mirrored). 

I am just interested in a design that allows rbd-mirror to not exceed its target memory but also increase its replay speed by being able to prefetch more data if it has the capacity.

trociny commented:
@dillaman I don't see how your view contradict with the current implementation. Right now, when a `JournalPlayer` add its cache to the cache manager, it tells that it wants up to `2 * <splay width> * <object size> * <number of replayers>` bytes (`max_size`). The cache manager "gives" as much as it can. The `JournalPlayer` checks how much it actually obtained before creating a new `ObjectPlayer` to set its `max_fetch_bytes`.

If in the future we want more memory to pre-fetch additional objects, we just update the `max_size` calculation, and then checking how much the cache manager actually gives, we can decide how many object players we actually want.

As I said your note about an `ObjectPlayer` created with high `max_fetch_bytes` and not doing immediate forward may be valid, but this can be resolved by just passing the cache handler (or something like this) to the `ObjectPlayer` so it could always calculate actual `max_fetch_bytes`, before every rados read. Though to me it seems not necessary.

To me the real issues with my implementation  is what I mentioned earlier: we can't call `Manager::insert`` to add the cache from the `JournalReplayer` without additional protection (serialization), and we need to make sure the manager ran rebalance after adding the cache and before we call `pri_cache->get_cache_bytes` to calculate `max_fetch_bytes` (otherwise we will always get 0).

I am going to update my PR the way I see it, so it might be easier to discuss.

trociny commented:
@dillaman how do you like this version?

When testing I stepped on another `PriorityCache` limitation: perf counters are not recycled and have rather limited  slots reserved, 16384, which is for about 1200 images. So I think I would need to implement  the counters recycling and also increase the reserved slots size. Alternatively  we could make `PriorityCache` perf counters optional and disable in our case.

Another question is what `max_fetch_bytes` value should be used when the auto-tune is disabled? Right now it allows to fetch the whole object, which is probably not a good idea. We could hardcode some value or just reuse `rbd_mirror_journal_max_fetch_bytes ` (which you wanted to remove).

dillaman commented on src/journal/JournalPlayer.cc:
```diff
@@ -800,4 +828,43 @@ void JournalPlayer::notify_complete(int r) {
+      return;
+    }
+  } else {
+    min_bytes /= 2;
```
What's the purpose behind this? Just so it creates a new minimum when not prefetching?

trociny replied:
The idea is that when we are short on memory it is preferable not to start a new JournalPlayer instead of shutting down the already running one. So on the start the minimum required is two times higher than when it is already running.

Still, I worry a little about a potential effect, when many images are replaying, no new ones are started and then by some reason the limit is reached. Then all replayers will be stopped simultaneously (because the balancer will reserve equally not enough space for them all). After this they all will be restarted and only part of them (the first ones) will be able to start, so it will functioning, but this shutdown looks like undesirable.

dillaman replied:
Is there a way that we can have it fail-fast? I agree it would be awkward if suddenly previously running replication was stopped just because a new image was added. That's kinda why I was wondering about adding the concept of reserving a minimum size from the cache for each ```JournalPlayer``` upon start-up. A player might be given more than that minimum amount if there is space but it will never be given less (maybe ```rbd_mirror_journal_max_fetch_bytes``` just becomes ```rbd_mirror_journal_min_fetch_bytes```?).

trociny replied:
To resolve this potential issue we could add some randomization:
- e.g. min_bytes instead of dividing by 2 could use some randomized factor so only part of the images are likely to reach `min_bytes` and would be stopped.
- or if min_bytes is reached we stop only with a probability 0.5 or something like this.

dillaman replied:
What about moving the ```m_cache_manager_handler``` call to ```prefetch``` and allow ```register_cache``` to return ```-ENOMEM``` if there is no more space for even the min fetch bytes? This error would then get bubbled up to the replay handler.

trociny commented:
@dillaman updated.

I did not move `m_cache_manager_handler` call to prefetch as you had suggested, as it is not trivial to make `register_cache` to return `-ENOMEM`. Taking the current `PriorityCache` implementation it is difficult to tell in advance if we have enough space to register yet another cache.  For this we need to actually register it and wait for rebalance to run and see the result. But as you mentioned this might cause that due to the newly added cache older caches will not obtain enough memory after rebalance and will be stopped. 

To resolve this problem I split the pri cache into two chunks: base and extra. The extra chunk always has the lowest priority, while the base (small) chunk has the lowest priority only at the moment of registration and the priority becomes higher after every rebalance.  This way when a new cache is added it will get some bytes only after all existing caches got some bytes for base chunk and there are still some free bytes.

Also I added some randomization when we check if the limit is reached, so we don't have the effect when many (all) images got not enough bytes and are stopped. With the randomization it is more likely that only some of them reach the limit first, will be stopped and then more space will be freed for other images.

And I made pri cache perf counters optional.
